### ISSUE TITLE
Updating a module parameter in-place with a different shape does not update the tensor shape in `state_dict`

### ISSUE THREAD

**Issue by fxmarty-amd** [pytorch/pytorch#137710]
### 🐛 Describe the bug

Hi,

I came across some code that updates "in place" a tensor, and not only updates its values but also its shape. Putting aside the choice of doing this, it seem to actually trigger a bug in pytorch:

```python
import torch.nn as nn

class MyModel(nn.Module):
    def __init__(self):
        super().__init__()
        
        self.register_buffer("my_val", torch.tensor(0.0))
    
    def forward(self, x):
        return x


model = MyModel()

state_dict = model.state_dict()

print("original model.my_val", model.my_val)
print("original data_ptr", model.my_val.data_ptr(), model.my_val.storage_offset())
print("original state_dict data_ptr", state_dict["my_val"].data_ptr(), state_dict["my_val"].storage_offset())


new_val = torch.rand(10)
model.my_val.resize_(new_val.shape)
model.my_val.copy_(new_val)

print("new model.my_val", model.my_val)
print("new model.my_val ptr", model.my_val.data_ptr(), model.my_val.storage_offset())

print("new data_ptr in state_dict", state_dict["my_val"].data_ptr(), state_dict["my_val"].storage_offset())
print("new state_dict", state_dict)
```

printing
```
original model.my_val tensor(0.)
original data_ptr 140491520 0
original state_dict data_ptr 140491520 0

new model.my_val tensor([0.8872, 0.1449, 0.4176, 0.0272, 0.7496, 0.4110, 0.6240, 0.8051, 0.8486,
        0.1659])
new model.my_val ptr 130642624 0

new data_ptr in state_dict 130642624 0
new state_dict OrderedDict([('my_val', tensor(0.8872))])
```

As you can note, the tensor in the state_dict shares the new pointer `130642624` but its shape is not updated after the `resize_` and `copy_` operations, hence the value in the buffer is only `tensor(0.8872)` (the first value in the larger tensor that was allocated).

This seems like a bug to me, although doing such operations is very much pushing boundaries of what should be legal. In my opinion, an error should be raised or the behavior should be as expected (update not only the pointer, but also the shape).

Thank you!

Reference: https://pytorch.org/docs/stable/generated/torch.Tensor.resize_.html

### Versions

```
Collecting environment information...
PyTorch version: 2.4.0+cpu
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: Ubuntu 22.04.4 LTS (x86_64)
GCC version: (Ubuntu 11.4.0-1ubuntu1~22.04) 11.4.0
Clang version: 14.0.0-1ubuntu1.1
CMake version: version 3.22.1
Libc version: glibc-2.35

Python version: 3.10.14 (main, May  6 2024, 19:42:50) [GCC 11.2.0] (64-bit runtime)
Python platform: Linux-5.15.153.1-microsoft-standard-WSL2-x86_64-with-glibc2.35
Is CUDA available: False
CUDA runtime version: No CUDA
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:                       x86_64
CPU op-mode(s):                     32-bit, 64-bit
Address sizes:                      48 bits physical, 48 bits virtual
Byte Order:                         Little Endian
CPU(s):                             16
On-line CPU(s) list:                0-15
Vendor ID:                          AuthenticAMD
Model name:                         AMD Ryzen 7 PRO 7840U w/ Radeon 780M Graphics
CPU family:                         25
Model:                              116
Thread(s) per core:                 2
Core(s) per socket:                 8
Socket(s):                          1
Stepping:                           1
BogoMIPS:                           6587.69
Flags:                              fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ht syscall nx mmxext fxsr_opt pdpe1gb rdtscp lm constant_tsc rep_good nopl tsc_reliable nonstop_tsc cpuid extd_apicid pni pclmulqdq ssse3 fma cx16 sse4_1 sse4_2 movbe popcnt aes xsave avx f16c rdrand hypervisor lahf_lm cmp_legacy svm cr8_legacy abm sse4a misalignsse 3dnowprefetch osvw topoext perfctr_core ssbd ibrs ibpb stibp vmmcall fsgsbase bmi1 avx2 smep bmi2 erms invpcid avx512f avx512dq rdseed adx smap avx512ifma clflushopt clwb avx512cd sha_ni avx512bw avx512vl xsaveopt xsavec xgetbv1 xsaves avx512_bf16 clzero xsaveerptr arat npt nrip_save tsc_scale vmcb_clean flushbyasid decodeassists pausefilter pfthreshold v_vmsave_vmload avx512vbmi umip avx512_vbmi2 gfni vaes vpclmulqdq avx512_vnni avx512_bitalg avx512_vpopcntdq rdpid fsrm
Virtualization:                     AMD-V
Hypervisor vendor:                  Microsoft
Virtualization type:                full
L1d cache:                          256 KiB (8 instances)
L1i cache:                          256 KiB (8 instances)
L2 cache:                           8 MiB (8 instances)
L3 cache:                           16 MiB (1 instance)
Vulnerability Gather data sampling: Not affected
Vulnerability Itlb multihit:        Not affected
Vulnerability L1tf:                 Not affected
Vulnerability Mds:                  Not affected
Vulnerability Meltdown:             Not affected
Vulnerability Mmio stale data:      Not affected
Vulnerability Retbleed:             Not affected
Vulnerability Spec rstack overflow: Mitigation; safe RET
Vulnerability Spec store bypass:    Mitigation; Speculative Store Bypass disabled via prctl and seccomp
Vulnerability Spectre v1:           Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:           Mitigation; Retpolines, IBPB conditional, IBRS_FW, STIBP conditional, RSB filling, PBRSB-eIBRS Not affected
Vulnerability Srbds:                Not affected
Vulnerability Tsx async abort:      Not affected

Versions of relevant libraries:
[pip3] mypy==1.6.1
[pip3] mypy-extensions==1.0.0
[pip3] numpy==1.26.4
[pip3] onnx==1.16.2
[pip3] onnxoptimizer==0.3.13
[pip3] onnxruntime==1.18.1
[pip3] onnxruntime_extensions==0.12.0
[pip3] onnxsim==0.4.36
[pip3] torch==2.4.0+cpu
[pip3] torchaudio==2.4.0+cpu
[pip3] torchvision==0.19.0+cpu
[conda] numpy                     1.26.4                   pypi_0    pypi
[conda] torch                     2.4.0+cpu                pypi_0    pypi
[conda] torchaudio                2.4.0+cpu                pypi_0    pypi
[conda] torchvision               0.19.0+cpu               pypi_0    pypi
```

cc @albanD @mruberry @jbschlosser @walterddr @mikaylagawarecki

**Comment by ezyang** [comment]
The proximal cause for this is that the tensor in the state dict and the tensor you resized are *different* tensors, and so metadata is not propagated. It's not entirely clear to me if them being different tensors is expected, but certainly given that they're different tensors this is expected behavior.

**Comment by fxmarty-amd** [comment]
The data pointer is still updated in the state dict, so surely they are not completely different tensors (aka, a clone)?

**Comment by BowenBao** [comment]
Indeed they are different tensors, which are created from `.detach()`. It looks like for OP setting `keep_vars` to True provides desired behavior.

https://github.com/pytorch/pytorch/blob/267f82b860ccdf32df3dcb92e2435b64ba0f117a/torch/nn/modules/module.py#L2173-L2176

```python
...
state_dict = model.state_dict(keep_vars=True)
...
```

```
original model.my_val tensor(0.)
original data_ptr 100167535629696 0
original state_dict data_ptr 100167535629696 0
new model.my_val tensor([0.5194, 0.4855, 0.9020, 0.2864, 0.2747, 0.6142, 0.3327, 0.0046, 0.7218,
        0.1832])
new model.my_val ptr 100167537183296 0
new data_ptr in state_dict 100167537183296 0
new state_dict OrderedDict([('my_val', tensor([0.5194, 0.4855, 0.9020, 0.2864, 0.2747, 0.6142, 0.3327, 0.0046, 0.7218,
        0.1832]))])
```

**Comment by ezyang** [comment]
ok, all sounds expected. Sorry but we can't fix this otherwise lol
